[interfaces/json-rpc] Allow additional properties in Files API calls

finkleandeinhorn opened the pull request:
## Description
This request will allow `ListItem` properties, set via `xbmcgui.ListItem.setProperty(...)`, to be retrieved via the `Files.GetDirectory` and `Files.GetFileDetails` JSON-RPC methods, by passing a list of the desired properties as the `additionalProperties` method parameter.

## Motivation and Context
The reasoning behind this change, as well as detailed discussion, can be found here: https://forum.kodi.tv/showthread.php?tid=355209

## How Has This Been Tested?
Also detailed in the thread linked above; this has been tested by running `Files.GetDirectory` calls against plugin directories that are known to set custom `ListItem` properties not typically available using the current implementation.

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

Lunatixz commented:
ping @ronie

enen92 commented:
@Tolriq can you also take a look?

enen92 reviewed (changes requested):
At least JSON-RPC minor version has to be bumped in `xbmc/interfaces/json-rpc/schema/version.txt `

ronie commented:
we have at least one team member who knows all about json-rpc.
that person, however, is not me ;-)

as i already mentioned on the forum: 'my code example is very likely not how additionalProperties should be implemented'

Tolriq commented:
Don't have time to fully explain but the implementation is wrong and does not really correspond to the description and need :)

Changes like https://github.com/xbmc/xbmc/pull/18106/files#diff-4252ea25ea01026195fc9eeaaec2a806R643 means that the requester can list the additionalProperties he wants, but code does not do that.

And https://github.com/xbmc/xbmc/pull/18106/files#diff-1bb266d43e6d1a19799e53fcf6371260R1597 says that returned field is a string when it's an array.

IMO "additional" or "additionalProperties" should be another normal property and handled like that.

finkleandeinhorn commented:
Actually, not quite.

First of all, [this line](https://github.com/xbmc/xbmc/pull/18106/files#diff-4252ea25ea01026195fc9eeaaec2a806R643) defines a separate parameter (as you mention in your final sentence, which takes in an array of strings, and the added code in `FileItemHandler` *does* actually handle that parameter properly.

Secondly, [this line](https://github.com/xbmc/xbmc/pull/18106/files#diff-1bb266d43e6d1a19799e53fcf6371260R1597) says that, in the returned response, additional properties beyond those that are defined in the schema can be included, *as long as* they are a string.

I'm absolutely open to changing the implementation, but so far I'm failing to see how this is *in*correct, as it actually *does* function the way it's described, has been tested to be fully working, and *shouldn't* affect any parts of the API besides the two mentioned methods.

finkleandeinhorn commented:
@enen92 Will do!

finkleandeinhorn commented:
Also, as a short explanation, it was @ronie's [initial inplementation](https://forum.kodi.tv/showthread.php?tid=355209&pid=2959200#pid2959200) which first used `additionalProperties` as a separate parameter, as well as the logic to handle them...

But the resulting values would get invalidated out of the response, because the set of valid properties didn't include arbitrary strings (which were technically already handled in the code, as I outline [here](https://forum.kodi.tv/showthread.php?tid=355209&pid=2959150#pid2959150)). Allowing additional string properties in the response makes this possible.

Tolriq commented:
Yes read a little too fast late night, the issue is that you mix a new param additionalProperties with an actual logic when this name  already have a specific different meaning and usage.
And that the returned values are randomly added as new fields in the normal answer making it impossible for any automapped parser (Main advantage of JSON schema ;)) to handle the data correctly. So I'm sorry but this is not exactly the most logical solution from general JSON API surface and current Kodi JSON implementation and naming.

The question is does it makes sense to request specific named additionalProperties when those are by definition unknown as undefined in the schema.

For me this need is the exact same need as uniqueId, you request that and it returns the array of unknown mappings.

Take the uniqueID code replace with a name that does not collide with additionalProperties (as it's well known for a specific purpose), and you have the proper solution to this issue that is consistent with how the JSON schema works.

DaveTBlake commented:
>At least JSON-RPC minor version has to be bumped in xbmc/interfaces/json-rpc/schema/version.txt 

Unfortunately inaccurate. When making a non-breaking change to the JSON API schema such as adding parameters, the minor number needs to be bumped (and patch set to 0) in `version.txt` https://github.com/xbmc/xbmc/blob/master/xbmc/interfaces/json-rpc/schema/version.txt . The schema documents this, see https://github.com/xbmc/xbmc/blob/master/xbmc/interfaces/json-rpc/schema/methods.json#L33

But while that is easily fixed, more fundamentally I also don't think this addition has been implemented correctly at all. It is probably frustrating to hear, as you have taken ronie's suggestion and run with it and it gives the results you want, but neither makes this the correct approach. We have to live with any JSON API changes for a long time to come, unable to do things because of a mess of breaking changes it brings, so better to get this right. @Tolriq is pointing you in the right direction.

finkleandeinhorn commented:
I'll continue down that road and see where we end up 😁

finkleandeinhorn commented:
@Tolriq, that implementation does sound like it would make more sense, but a cursory browse of the code makes me a bit more confused...

`uniqueid` appears to be a map that is stored in the items `VideoInfoTag`, but I don't see any method to retrieve a map of the assigned properties of the item, and I'm having trouble figuring out which member would be used to store such a map, if it exists.

Tolriq commented:
This is the problem of trying to mix GUI and non GUI stuff.

Anyway from a quick look it seems that the properties are stored in m_mapProperties of CGUIListItem 

So something like 

```
  for (const auto& it : m_mapProperties)
  {
    value["listItemProperties"][it.first] = it.second;
  }
```

Added to ```void CFileItem::Serialize(CVariant& value) const``` (https://github.com/xbmc/xbmc/blob/master/xbmc/FileItem.cpp#L647)

And adding listItemProperties to the list of properties as the uniqueId should work. (Completely untested)

finkleandeinhorn commented:
@Tolriq Hmm, my first pass at that didn't quite work out...

I've added lines similar to these: https://github.com/xbmc/xbmc/blob/7493cd936692a0f420aa39e79ad66cd13d7863ee/xbmc/interfaces/json-rpc/schema/types.json#L443-450, and referenced them as the type for the new property I added to `List.Item.Base`, as well as `List.Fields All` and `List.Fields.Files`, similar to how [`uniqueid` is declared](https://github.com/xbmc/xbmc/blob/7493cd936692a0f420aa39e79ad66cd13d7863ee/xbmc/interfaces/json-rpc/schema/types.json#L443)...

What am I missing here?

finkleandeinhorn commented:
Actually, I believe I've gotten it (seems I just had a typo). The current implementation doesn't allow *sending* arbitrary values in the query (only `listitemproperties`), but returns a map of what appears to be all of the properties associated with that item, as `listitemproperties`, in the response.

Tolriq commented:
Looks better :)

You need to remove the no more needed changes in FileItemHandler.cpp

About version it's the minor and not revision that needs change so 11.12.0 is the correct version.

For the m_mapProperties serializing you may need to add a null check before don't know if it can be null or not

finkleandeinhorn commented:
Ah, I didn't realize that old code was still in there 😅 I fixed those things up (with what I think should be an acceptable null check?), and took the liberty of squashing the whole mess for you guys 😁

Tolriq commented:
Thanks, sorry I forget to say that you need to remove also Item.Properties.Set as it's not used.

And I'm wondering about the Item.Properties naming. (But I'm really bad at naming stuff) Maybe ListItem.Property ? @DaveTBlake 

And to finish since I lack time, can you please give example output when requesting this property for addons path that should have some data in is as for your previous tests + some for normal directory listing that should probably return an empty map.

finkleandeinhorn commented:
@Tolriq I mainly chose `Item.Properties` because there don't seem to be any *other* `ListItem.*` types 🤷🏾‍♂️

I'll definitely take out `Item.Properties.Set`, spin up a new build, and get some example usage logged out.

Being that this is being requested towards the Matrix branch, is there any possibility of a backport in time for an 18.8 release as well?

Tolriq commented:
Yes there's no other ListItem types and List.Item is something else. But the property is called listitemproperties as you picked the mostly random name I proposed :p

Let's see what others think about it.

About 18.8 there won't be new features merged. Specially JSON as versioning properly is impossible.

finkleandeinhorn commented:
Well, the build failed 😅 It's because that apparently isn't the proper way to handle a null check in C++... I'll get that fixed.

finkleandeinhorn commented:
So, while trying to find some example queries for plugins that *don't* have any `ListItem` properties to return, I've found that I can't really find any add-ons that *don't* use them at least a little bit.

With that being said, I've decided to simply give a few of the queries I've tested, and an example file from the response of each.

A case which retrieves only `listitemproperties`: `{"jsonrpc":"2.0","id":27,"method":"Files.GetDirectory","params":{"directory":"plugin://plugin.video.themoviedb.helper/?info=popular&nextpage=True&type=tv","media":"video","properties":["listitemproperties"]}}`
```json
{
    "file": "plugin://plugin.video.themoviedb.helper/?info=details&type=…",
    "filetype": "directory",
    "label": "Dark",
    "listitemproperties": {
        "container.info": "popular",
        "container.nextpage": "True",
        "container.type": "tv",
        "dbtype": "tvshow",
        "item.info": "details",
        "item.nextpage": "True",
        "item.tmdb_id": "70523",
        "item.type": "tv",
        "numitems.dbid": "0",
        "numitems.tmdb": "20",
        "tmdb_id": "70523",
        "tmdb_rating": "8.5",
        "tmdb_votes": "2,032",
        "widget": "TMDb Helper"
    },
    "type": "tvshow"
}
```

A case that retrieves `listitemproperties`, as well as another property: `{"jsonrpc":"2.0","id":34,"method":"Files.GetDirectory","params":{"directory":"plugin://plugin.video.twitch/?mode=featured_streams","media":"video","properties":["plot","listitemproperties"]}}`
```json
{
    "file": "plugin://plugin.video.twitch/?mode=play&channel_id=12327567…",
    "filetype": "file",
    "label": "AmazonMusic - 6/29 (4PM PT) - Noah Reid Live!",
    "listitemproperties": {
        "contextmenuaction(0)": "RunScript(special://home/addons/plugin.video.twitch/resourc…",
        "contextmenuaction(1)": "Container.Update(plugin://plugin.video.twitch/?mode=channel…",
        "contextmenuaction(2)": "Container.Update(plugin://plugin.video.twitch/?mode=game_li…",
        "contextmenuaction(3)": "RunPlugin(plugin://plugin.video.twitch/?mode=edit_blacklist…",
        "contextmenuaction(4)": "RunPlugin(plugin://plugin.video.twitch/?mode=edit_blacklist…",
        "contextmenuaction(5)": "RunPlugin(plugin://plugin.video.twitch/?mode=edit_qualities…",
        "contextmenuaction(6)": "RunPlugin(plugin://plugin.video.twitch/?mode=play&channel_i…",
        "contextmenulabel(0)": "Refresh",
        "contextmenulabel(1)": "Go to [COLOR=white][B]AmazonMusic[/B][/COLOR]",
        "contextmenulabel(2)": "Go to [COLOR=white][B]Music & Performing Arts[/B][/COLOR]",
        "contextmenulabel(3)": "Blacklist [COLOR=white][B]AmazonMusic[/B][/COLOR]",
        "contextmenulabel(4)": "Blacklist [COLOR=white][B]Music & Performing Arts[/B][/COLO…",
        "contextmenulabel(5)": "Set default quality",
        "contextmenulabel(6)": "Play (Choose Quality)",
        "isplayable": "true"
    },
    "plot": "6/29 (4PM PT) - Noah Reid Live!\r\nGame: Music & Performing…",
    "type": "unknown"
}
```

And of course, a case which *doesn't* request `listitemproperties`: `{"jsonrpc":"2.0","id":38,"method":"Files.GetDirectory","params":{"directory":"plugin://script.embuary.info/movie/top_rated","media":"video","properties":["genre","director"]}}`
```json
{
    "director": [],
    "file": "plugin://script.embuary.info/info/movie/tmdb/496243",
    "filetype": "file",
    "genre": [],
    "label": "Parasite",
    "type": "movie"
}
```

I'm going to go ahead and squash my last couple commits as well, so that (unless it's deemed that I still need to change the name of `Item.Properties`) this is ready to merge.

finkleandeinhorn reviewed (commented):
This wasn't it! 😅

finkleandeinhorn commented:
Why is that build failing? 🤔

fuzzard commented:
You must have done a force push whilst it was building.

Jenkins build this please

finkleandeinhorn commented:
@fuzzard Also, this is my first time with Jenkins... Why doesn't he build EXEs or installers?

fuzzard commented:
Artifacts are only built for some platforms (android/apple). As i understand it, its more a data storage/bandwidth decision.

Tolriq commented:
Thanks for the examples.

I'm really amazed to see 
```
 "listitemproperties": {
        "contextmenuaction(0)": "RunScript(special://home/addons/plugin.video.twitch/resourc…",
        "contextmenuaction(1)": "Container.Update(plugin://plugin.video.twitch/?mode=channel…",
        "contextmenuaction(2)": "Container.Update(plugin://plugin.video.twitch/?mode=game_li…",
        "contextmenuaction(3)": "RunPlugin(plugin://plugin.video.twitch/?mode=edit_blacklist…",
        "contextmenuaction(4)": "RunPlugin(plugin://plugin.video.twitch/?mode=edit_blacklist…",
        "contextmenuaction(5)": "RunPlugin(plugin://plugin.video.twitch/?mode=edit_qualities…",
        "contextmenuaction(6)": "RunPlugin(plugin://plugin.video.twitch/?mode=play&channel_i…",
        "contextmenulabel(0)": "Refresh",
        "contextmenulabel(1)": "Go to [COLOR=white][B]AmazonMusic[/B][/COLOR]",
        "contextmenulabel(2)": "Go to [COLOR=white][B]Music & Performing Arts[/B][/COLOR]",
        "contextmenulabel(3)": "Blacklist [COLOR=white][B]AmazonMusic[/B][/COLOR]",
        "contextmenulabel(4)": "Blacklist [COLOR=white][B]Music & Performing Arts[/B][/COLO…",
        "contextmenulabel(5)": "Set default quality",
        "contextmenulabel(6)": "Play (Choose Quality)",
        "isplayable": "true"
    }
```

Is this the official way for plugins to push context menu @enen92 or if someone else know? Exposing the context menus in Yatse was asked quite a few times, so if this is the official / only way it means it will now be possible as a side effect of this PR.

DaveTBlake commented:
Naming things, my suggestion would be
```
"Item.CustomProperties": {
    "type": "object",
    "additionalProperties": { "$ref": "Global.String.NotEmpty" }
  },
```
and
```
"customproperties": { "$ref": "Item.CustomProperties" }
```

enen92 commented:
@Tolriq yes, context menu items in python are added to listitems using `listitem.addContextMenuItems({("label", "action")...}` (see https://codedocs.xyz/xbmc/xbmc/group__python__xbmcgui__listitem.html#ga76a5f710e5c08dc2033a09e23a1a78c2) which receives a list containing tuples of ("Label", "Builtin-function"). Internally this is just equivalent of adding individual properties to the item object referencing an incremental id: `contextmenulabel(id)` and `contextmenuaction(id)` (see https://github.com/xbmc/xbmc/blob/fd8a20d2b7b2b510bb599cd8225ba5ea2a949d09/xbmc/interfaces/legacy/ListItem.cpp#L769-L770).

Thanks for moving this PR forward and guiding the submitter.

enen92 commented:
Btw, when adding subtitles to a ListItem (`listitem.setSubtitles(...)`) Kodi also sets them as properties: https://github.com/xbmc/xbmc/blob/fd8a20d2b7b2b510bb599cd8225ba5ea2a949d09/xbmc/interfaces/legacy/ListItem.cpp#L780-L781

It's unclear to me if such things should be available to external consumers. Kodi (and addons) use properties sometimes as a mere vehicle of information between components due to the fact `CFileItems` are passed all the way through the chain. I leave that to anyone more interested than me in the json-rpc to decide. For example, if you consider `inputstream` based addons, by exposing properties to outsiders you are automatically exposing sensitive information such as license servers, license keys, etc

enen92 reviewed (commented):
My main concern with this is the possible exposure of "sensitive" information. I'd probably consider to create a vector of strings in FileItem `Serialize()` that could be added as a blacklist of properties to be hidden from consumers. That vector could be easily extended if more sensitive properties are added and used by kodi core. The only candidate I see atm are the properties started by "inputstream*".

finkleandeinhorn commented:
@Tolriq They were very interesting to see there, yeah. Another one that surprised me (which had almost all of them filled by TMDbHelper's service monitor) was a TMDbHelper item with 640 properties!

@DaveTBlake Works for me 😁

@enen92 Unless I'm mistaken, the `inputstream` add-ons also don't expose any lists? They don't appear to have any directories these calls would apply to. Remember that this is still `Files.etDirectory` 😅

enen92 commented:
@finkleandeinhorn I mean plugins that rely on inputstream addons for playback. If the plugin happens to provide a directory with final items (e.g. listitems.path is the url of the stream and the required inputstream properties are set on that list), they will be exposed to JSON and may include sensitive information. Btw this was supposed to be just a comment, maybe I made a mistake and requested changes instead.

finkleandeinhorn commented:
Ah, I see... Would YouTube be a potential candidate to test? I don't know of many add-ons that use it, and are Matrix-compatible 🤔

I can, of course, run a build of Leia with these changes to test with, but I had intended on simply doing that (if not for anything but personal use) once the changes are finalized. It's such a hassle to switch workspaces and rebuild, when I'm going to have to return here anyways 😅

finkleandeinhorn commented:
With Netflix:
`{"jsonrpc":"2.0","id":11,"method":"Files.GetDirectory","params":{"directory":"plugin://plugin.video.netflix/directory/video_list/currentTitles/trendingNow/","media":"video","properties":["listitemproperties"]}}`
```json
{
    "file": "plugin://plugin.video.netflix/directory/show/70143860/?prof…",
    "filetype": "directory",
    "label": "The Vampire Diaries",
    "listitemproperties": {
        "contextmenuaction(0)": "RunPlugin(plugin://plugin.video.netflix/action/my_list/add/…",
        "contextmenuaction(1)": "RunPlugin(plugin://plugin.video.netflix/action/trailer/show…",
        "contextmenuaction(2)": "RunPlugin(plugin://plugin.video.netflix/action/rate_thumb/s…",
        "contextmenuaction(3)": "RunPlugin(plugin://plugin.video.netflix/library/export/show…",
        "contextmenulabel(0)": "Add to 'My list'",
        "contextmenulabel(1)": "Show trailers & more",
        "contextmenulabel(2)": "Rate on Netflix",
        "contextmenulabel(3)": "Export to library",
        "isfolder": "True"
    },
    "type": "tvshow"
}
```

With YouTube:
`{"jsonrpc":"2.0","id":7,"method":"Files.GetDirectory","params":{"directory":"plugin://plugin.video.youtube/special/popular_right_now/","media":"video","properties":["listitemproperties"]}}`
```json
{
    "file": "plugin://plugin.video.youtube/play/?video_id=h0U2QUGKbSE",
    "filetype": "file",
    "label": "Kanye West – Wash Us In The Blood feat. Travis Scott (Offic…",
    "listitemproperties": {
        "channel_id": "UCs6eXM7s8Vl5WcECcRHc2qQ",
        "contextmenuaction(0)": "Container.Refresh",
        "contextmenuaction(1)": "Action(Queue)",
        "contextmenuaction(2)": "Container.Update(plugin://plugin.video.youtube/channel/UCs6…",
        "contextmenuaction(3)": "RunPlugin(plugin://plugin.video.youtube/video/more/?video_i…",
        "contextmenuaction(4)": "RunPlugin(plugin://plugin.video.youtube/play/?video_id=h0U2…",
        "contextmenuaction(5)": "RunPlugin(plugin://plugin.video.youtube/play/?video_id=h0U2…",
        "contextmenuaction(6)": "RunPlugin(plugin://plugin.video.youtube/play/?video_id=h0U2…",
        "contextmenulabel(0)": "Refresh",
        "contextmenulabel(1)": "Queue video",
        "contextmenulabel(2)": "Go to [B]Kanye West[/B]",
        "contextmenulabel(3)": "More...",
        "contextmenulabel(4)": "Play with subtitles",
        "contextmenulabel(5)": "Play audio only",
        "contextmenulabel(6)": "Play (Ask for quality)",
        "isplayable": "true",
        "publishedlocal": "2020-06-30 07:00:11",
        "publishedsince": "Today at 07:00:11 AM"
    },
    "type": "unknown"
}
```

And even a live link from YouTube: `{"jsonrpc":"2.0","id":16,"method":"Files.GetDirectory","params":{"directory":"plugin://plugin.video.youtube/special/live/","media":"music","properties":["listitemproperties"]}}`
```json
{
    "file": "plugin://plugin.video.youtube/play/?video_id=VefL3xoCS5A",
    "filetype": "file",
    "label": "BRIGHTON vs MANCHESTER UNITED With MARK GOLDBRIDGE LIVE",
    "listitemproperties": {
        "channel_id": "UCMmVPVb0BwSIOWVeDwlPocQ",
        "contextmenuaction(0)": "Container.Refresh",
        "contextmenuaction(1)": "Action(Queue)",
        "contextmenuaction(2)": "Container.Update(plugin://plugin.video.youtube/channel/UCMm…",
        "contextmenuaction(3)": "RunPlugin(plugin://plugin.video.youtube/video/more/?video_i…",
        "contextmenuaction(4)": "RunPlugin(plugin://plugin.video.youtube/play/?video_id=VefL…",
        "contextmenulabel(0)": "Refresh",
        "contextmenulabel(1)": "Queue video",
        "contextmenulabel(2)": "Go to [B]The United Stand[/B]",
        "contextmenulabel(3)": "More...",
        "contextmenulabel(4)": "Play (Ask for quality)",
        "isplayable": "true",
        "publishedsince": "Live"
    },
    "type": "unknown"
}
```

So unless the `contextmenu` properties are a security issue (I'd think not, and that this actually opens up even more potential legitimate usages), it appears that even `inputstream` based add-ons are going to be fine here 😁

finkleandeinhorn commented:
Jenkins, could you build this one again? 😁

enen92 commented:
Yes, those add-ons are properly implemented as they route the playable items to specific plugin:// paths (that are not directories - isFolder=false). When called those plugin:// paths set the playable path (dynpath) using `setresolvedurl`. The situation I pointed out is only possible if the directory itself contain items with the path already already pointing to the final stream URL. It is unlikely but...can happen. Thanks for taking the time to run some tests though.
I would not worry to much about this possible problem at the moment.

fuzzard commented:
Jenkins build this please

enen92 reviewed (approved):
LGTM. Regarding the exposure of sensitive information I would not worry too much with it at this point. It's highly unlikely (see discussion) and we can always improve it later with some blacklist or similar approach.

finkleandeinhorn commented:
@Tolriq 
> About 18.8 there won't be new features merged. Specially JSON as versioning properly is impossible.

It appears the current JSON version in Leia is `10.3.0`... would a minor bump to `10.4.0` not be acceptable? The API is a _hair_ different in Leia, but I could open another PR backporting the relevant changes. I would just really love to see this make it into Leia, as it seems Matrix is still a ways off from a fully stable release and/or wide adoption.

Tolriq commented:
Please post result of a test for a non plugin folder as asked earlier.

```{"jsonrpc":"2.0","id":34,"method":"Files.GetDirectory","params":{"directory":"/mnt/media/","media":"video","properties":["plot","listitemproperties"]}}```

@DaveTBlake will confirm but there's never API addition to backport, (for anything actually) but it's even more true for JSON.

10.4.0 already exist in Leia: https://github.com/xbmc/xbmc/pull/15869 how would a consumer know that 10.4.0 can represent 2 different things :) That's not how public API versioning works.

Since there's never new functions backport there's no major bump on new Kodi version branching leading to this situation that in the end makes those impossible.

DaveTBlake commented:
No backporting @finkleandeinhorn 
a) Generaly only fixes are backported, not new features like this
b) Versions API v10.4.0 and above exists on the Matrix branch with additions to the API that were not backported to v18, therefore unless all the steps before v11.0.0 were also backported there is no way to keep version consistency.

finkleandeinhorn commented:
@Tolriq @DaveTBlake Ah, I understand now.

However, using any non-plugin path wasn't successful in my tests. using a `videodb://` path returned an empty response (I assume you're to get at these with the `VideoLibrary` methods), and a normal folder path returned an "Invalid params" error, whether `special://`-formatted or not.

Tolriq commented:
Just add a source in Kodi and ask for that path, if the query does not work then you have maybe broken the feature ;) Hence the ask to test as I lack times to do it.

finkleandeinhorn commented:
@Tolriq I didn't realize I needed to add it as a source first 😅

`{"jsonrpc":"2.0","id":12,"method":"Files.GetDirectory","params":{"directory":"C:\\Users\\finkleandeinhorn\\Documents","media":"video","properties":["listitemproperties"]}}`
```json
{
    "file": "C:\\Users\\David\\Documents\\GitHub\\",
    "filetype": "directory",
    "label": "GitHub",
    "type": "unknown"
}
```

But... Should it *only* be returning directories? I've tested the same exact call in a fresh 18.7 install, and have the same behavior, but that call only returns directory (and archive) items, but no files at all. Is that a new issue? Or expected behavior?

Tolriq commented:
You are requesting media video so it returns only scrapped video and folders, use media : files to get all.

finkleandeinhorn commented:
@Tolriq Ah ok, so one like this:
`{"jsonrpc":"2.0","id":6,"method":"Files.GetDirectory","params":{"directory":"C:\\Users\\finkleandeinhorn\\Documents","media":"files","properties":["customproperties"]}}
`
```json
{
    "file": "C:\\Users\\finkleandeinhorn\\Documents\\Graph_and_Line_Paper.pdf",
    "filetype": "file",
    "label": "Graph_and_Line_Paper.pdf",
    "type": "unknown"
}
```

Tolriq commented:
Yep, so all working :) No crash and no null returned.

From technical POV this PR is now OK. 

For the risk of data exposure this is not my call (But I suppose data is already exposed internally to other addons so this is something larger to think about than this PR as @enen92 said).
If this can be an issue it should be discussed before merging to avoid breaking API change later.

finkleandeinhorn commented:
@DaveTBlake To clarify, what is the revision number in `version.txt` used for? There doesn't appear to be any "revision" version (like 10.3.1, for example?) that would be a conflict if this was merged into the Leia branch, and that would ensure that Leia users could take advantage of add-ons that use this feature, long before they may be able to effectively do so in Matrix.

I suppose I don't understand the reasoning behind **not** allowing this into Leia, as the changes are so minimal, and shouldn't affect any other parts of the API that I can tell. Even for users of the API without these changes, their existing API calls will be compatible with no changes (as omitting the added property doesn't cause any issues at all, as is the case with any of the pre-existing properties), and allow for further feature development.

If I'm simply barking up the wrong tree for that one, I'll drop it, but it doesn't seem to make sense to not at least backport this as a "revision".

fuzzard commented:
Stance is always, bugfixes only for released branches. It wont be put in to Leia.

Tolriq commented:
Revision is for bug fixes that does not change the API surface, minor is for API addition that does not break anything and major is for breaking change that require users to adapt.

You can't change definition of things just to fit your needs ;)

The JSON API definition is outside the Kodi versioning. People that want to use a feature request JSON version for that, because if you request the field on a Kodi version that does not support it the query does not work.

So one can not just check if the JSON version is > to 10.3.1 to know if a feature is present as 10.4.0 is superior but does not contain it. If you need to check both Kodi and JSON version then JSON version have failed it's purpose.

finkleandeinhorn commented:
@Tolriq Ok, understood. I guess once this is merged, it'll give me a good reason to start migrating more seriously towards Matrix 😅

finkleandeinhorn commented:
Were @Montellese and @DaveTBlake wanting to weigh in as well?

I've been using this in my own tests and it's working like a charm.

Montellese commented on xbmc/FileItem.cpp:
```diff
@@ -668,6 +668,10 @@ void CFileItem::Serialize(CVariant& value) const
+
+  if (!m_mapProperties.empty())
+    for (const auto& it : m_mapProperties)
+      value["customproperties"][it.first] = it.second;
```
It would probably be more efficient to first get a reference to `value["customproperties"]` and use that instead of having to implicitely look it up for every property.

finkleandeinhorn replied:
Could you offer guidance on that? I'm apparently not quite familiar enough with C++ to get my head around it 😅

Montellese replied:
```cpp
auto& customProperties = values["customproperties"];
for (const auto& prop : m_mapProperties)
  customProperties[prop.first] = prop.second;
```

finkleandeinhorn replied:
Using that code, I got the following errors on build:
```python
"C:\Users\finkleandeinhorn\Documents\GitHub\xbmc\kodi-build.x64\ALL_BUILD.vcxproj" (default target) (1) ->
"C:\Users\finkleandeinhorn\Documents\GitHub\xbmc\kodi-build.x64\kodi.vcxproj" (default target) (12) ->
"C:\Users\finkleandeinhorn\Documents\GitHub\xbmc\kodi-build.x64\libkodi.vcxproj" (default target) (14) ->
(ClCompile target) ->
  c:\users\finkleandeinhorn\documents\github\xbmc\xbmc\fileitem.cpp(675): error C2065: 'customProperties': undeclared identifier [
C:\Users\finkleandeinhorn\Documents\GitHub\xbmc\kodi-build.x64\libkodi.vcxproj]
```
Which is especially confusing to me, because it *has* been declared 🤔

DaveTBlake replied:
`auto& customProperties = value["customproperties"];`
value not values

finkleandeinhorn replied:
Yes, I caught that change myself. The error above is something else.

Should it maybe be `const auto& customProperties`?

DaveTBlake replied:
No `const` is wrong, you are changing the value. 
```
  auto& customProperties = value["customproperties"];
  for (const auto& prop : m_mapProperties)
    customProperties[prop.first] = prop.second;
```
complies for me (on VS2017). Commit the change and we will see what jenkins does.

Montellese reviewed (commented):
Ok with me apart from the proposed improvement.

DaveTBlake commented on xbmc/FileItem.cpp:
```diff
@@ -668,6 +668,11 @@ void CFileItem::Serialize(CVariant& value) const
   if (m_gameInfoTag)
     (*m_gameInfoTag).Serialize(value["gameInfoTag"]);
+
+  if (!m_mapProperties.empty())
+    auto& customProperties = value["customproperties"];
+    for (const auto& prop : m_mapProperties)
+      customProperties[prop.first] = prop.second;
```
Wrong logic, not what was suggested. 
Remove the `if` which is only controlling the variable declaration and not the `for` loop. Hence loop (which is always hit)  is trying to use a variable that is only sometimes defined.

finkleandeinhorn replied:
Ah, I see. I had added that in response to the desire for a "null check" before reading the array, so I figured it was still necessary.

Removed now 👌

finkleandeinhorn replied:
Or, of that one succeeds, then a *proper* conditional makes the most sense to me.

finkleandeinhorn replied:
Looks like we need Jenkins to build this again, please?

finkleandeinhorn replied:
@DaveTBlake Well, Jenkins succeeded now, so I'll try a local build to test the changes, but I think this is probably good now.

Montellese replied:
Please keep the `if` statement because otherwise you will always create an empty `customproperties` sub-object even if it is empty.

finkleandeinhorn replied:
Alright, I've added the conditional back, and tested in a local build. This seems to be working, both to retrieve properties that exist:
`{"jsonrpc":"2.0","id":15,"method":"Files.GetDirectory","params":{"directory":"plugin://plugin.video.themoviedb.helper/?info=trending_week&nextpage=True&type=movie","media":"video","properties":["customproperties"]}}`
```json
{
    "customproperties": {
        "container.info": "trending_week",
        "container.nextpage": "True",
        "container.type": "movie",
        "dbtype": "movie",
        "item.info": "details",
        "item.nextpage": "True",
        "item.tmdb_id": "531454",
        "item.type": "movie",
        "numitems.dbid": "0",
        "numitems.tmdb": "20",
        "tmdb_id": "531454",
        "tmdb_rating": "6.5",
        "tmdb_votes": "283",
        "widget": "TMDb Helper"
    },
    "file": "plugin://plugin.video.themoviedb.helper/?info=details&type=…",
    "filetype": "directory",
    "label": "Eurovision Song Contest: The Story of Fire Saga",
    "type": "movie"
}
```
and also holds true (not sending non-existent properties) when querying a path with no properties:
`{"jsonrpc":"2.0","id":17,"method":"Files.GetDirectory","params":{"directory":"C:\\Users\\finkleandeinhorn\\Documents","media":"files","properties":["customproperties"]}}`
```json
{
    "file": "C:\\Users\\David\\Documents\\blue-dark-yellow-abstract-arti…",
    "filetype": "file",
    "label": "blue-dark-yellow-abstract-artistic-4k-hw.jpg",
    "type": "unknown"
}
```

For me, everything seems in order here. I'm going to go ahead and squash these ridiculous commits 😅

finkleandeinhorn replied:
Seems we need Jenkins to try again since I force-pushed 😉

lrusak commented:
jenkins build this please

dmyoung9 committed:
[interfaces/json-rpc] Retrieve ListItem properties via Files API

finkleandeinhorn commented:
Jenkins build this please

DaveTBlake commented:
jenkins build this please

DaveTBlake commented:
OK, got there in the end :)

finkleandeinhorn commented:
Thanks for the merge (and all your patience 😅)!

After going ahead and playing with your more, I've come to understand why the versioning prevents it from moving to Leia... since the only way I would have to check is if the version is at least `11.12.0`. I would've loved to see it come to the public a bit sooner, but this is great 😁
